packages/kbn-health-gateway-server/src/kibana/routes/root.ts

```diff
@@ -6,164 +6,160 @@
-      }`
-    );
-  });
+  readonly method = 'GET';
+  readonly path = '/';
 
-  return responses;
-}
+  constructor(private kibanaConfig: KibanaConfig, private logger: Logger) {
+    this.handler = this.handler.bind(this);
 
-function mergeResponses(
-  responses: Array<PromiseFulfilledResult<Response> | PromiseRejectedResult>
-) {
-  const hasUnhealthyResponse = responses.some(isUnhealthyResponse);
+    return pick(this, ['method', 'path', 'handler']) as RootRoute;
+  }
 
-  return {
-    body: {}, // The control plane health check ignores the body, so we do the same
-    statusCode: hasUnhealthyResponse ? 503 : 200,
-  };
-}
+  async handler(request: Request, toolkit: ResponseToolkit): Promise<ResponseObject> {
+    const body = await this.poll();
+    const code = RootRoute.STATUS_CODE[body.status];
 
-function isUnhealthyResponse(response: PromiseFulfilledResult<Response> | PromiseRejectedResult) {
-  return (
-    response.status === 'rejected' || !(isSuccess(response.value) || isUnauthorized(response.value))
-  );
-}
+    this.logger.debug(`Returning ${code} response with body: ${JSON.stringify(body)}`);
 
-function isUnauthorized({ status, headers }: Response): boolean {
-  return status === 401 && headers.has('www-authenticate');
-}
+    return toolkit.response(body).type('application/json').code(code);
+  }
 
-function isSuccess({ status }: Response): boolean {
-  return (status >= 200 && status <= 299) || status === 302;
-}
+  private async poll(): Promise<RootRouteResponse> {
+    const hosts = await Promise.all(this.kibanaConfig.hosts.map(this.pollHost.bind(this)));
+    const statuses = chain(hosts).map('status').uniq().value();
+    const status = statuses.length <= 1 ? statuses[0] ?? 'healthy' : 'unhealthy';
 
-function generateAgentConfig(sslConfig: KibanaConfig['ssl']) {
-  const options: https.AgentOptions = {
-    ca: sslConfig.certificateAuthorities,
-    cert: sslConfig.certificate,
-  };
-
-  const verificationMode = sslConfig.verificationMode;
-  switch (verificationMode) {
-    case 'none':
-      options.rejectUnauthorized = false;
-      break;
-    case 'certificate':
-      options.rejectUnauthorized = true;
-      // by default, NodeJS is checking the server identify
-      options.checkServerIdentity = () => undefined;
-      break;
-    case 'full':
-      options.rejectUnauthorized = true;
-      break;
-    default:
-      throw new Error(`Unknown ssl verificationMode: ${verificationMode}`);
+    return {
+      status,
+      hosts,
+    };
   }
 
-  return options;
-}
+  private async pollHost(host: string): Promise<HostStatus> {
+    const url = `${host}${RootRoute.POLL_ROUTE}`;
+    this.logger.debug(`Requesting ${url}`);
 
-function configureFetch(kibanaConfig: KibanaConfig) {
-  let agent: https.Agent;
+    try {
+      const response = await this.fetch(url);
+      this.logger.debug(`Healthy response from ${url} with code ${response.status}`);
```

Comment:
Do we want to conditionally log healthy/unhealthy by checking `RootRoute.isHealthy(response)` here as well?

Reply:
Makes total sense. I've updated that.